gh-87389: avoid treating path as URI with netloc

nascheme opened the pull request:
I believe this is a more correct fix for gh-87389 and may fix security bugs in other applications using `urlunsplit()`.  We should not be confusing the `path` argument as a netloc, IMHO.  I fixed `http.server` by not using `urllib.parse` on the `path`.  For `http.server`, that part of the HTTP request is not treated as a full URL but instead a path + optional query + optional fragment.  So just parsing it "by hand" and not using `urllib.parse` avoids the bug.

nascheme committed:
wip: alternative fix for gh-87389

still needs test for urlunsplit() change

nascheme committed:
Don't strip slashes if not building relative URL.

nascheme committed:
Add unit test for urlunsplit relative URL case.

gpshead commented on Lib/urllib/parse.py:
```diff
@@ -497,9 +497,15 @@ def urlunsplit(components):
+    scheme, netloc, path, query, fragment, _coerce_result = (
                                           _coerce_args(*components))
-    if netloc or (scheme and scheme in uses_netloc and url[:2] != '//'):
+    if scheme in uses_netloc and path.startswith('//'):
```
in practice this means someone passed in invalid components, but I agree that this is good desirable logic assuming uses_netloc is correct.  looking at uses_netloc, it contains the protocols anyone cares about and critically contains the empty string so this alone would fix the reported issue given the current code.

nascheme replied:
I reverted the change to this line.  I'm not really sure what's it's trying to do so I decided better to not change it.  It seems like a weird edge case that doesn't make sense to me.  I.e. a scheme, not netloc and path that doesn't start with //.  E.g.
```
In [4]: urllib.parse.urlunsplit(('http', '', '/foo', '', ''))
Out[4]: 'http:///foo'
```
That doesn't seem right to me.  Maybe it makes sense to have scheme and not netloc for certain schemes?

gpshead replied:
I don't think urllib.parse's existing API implementations are great... there are cases we should error on or not allow or sanitize in things like unsplit and unparse but things probably depend on this.  reverting the change when there's a question is the right thing to do. :)   https://datatracker.ietf.org/doc/html/rfc3986 came about to specify URIs many years after this code was originally written.

gpshead commented on Lib/urllib/parse.py:
```diff
@@ -497,9 +497,15 @@ def urlunsplit(components):
+    scheme, netloc, path, query, fragment, _coerce_result = (
                                           _coerce_args(*components))
-    if netloc or (scheme and scheme in uses_netloc and url[:2] != '//'):
+    if scheme in uses_netloc and path.startswith('//'):
```
Reorder these test. `path.startswith('//')` is the fast operation, do it first. `scheme in uses_netloc` is potentially slow as uses_netloc is a list in suboptimal order rather than a set.

gpshead commented on Lib/http/server.py:
```diff
@@ -881,6 +878,17 @@ def guess_type(self, path):
         return 'application/octet-stream'
 
 
+def _add_trailing_slash(path):
```
I don't like this function. It duplicates what urllib.parse.urlsplit does and is putting logic around parsing URIs in a distant place far from that. It also omits other important behaviors or urlsplit such as control character removal. These are the kinds of things that should not be duplicated logic.

If this is going to exist, it should live in urllib/parse.py.  It should also indicate that it has a dual return value in its name. Returning None for a function claiming to modify something is unusual.  `_add_trailing_slash_or_none` would make it more obvious to someone reading code calling this that a falsey result is expected when no work was found to be done.

like urlsplit, this also deserves a `@functools.lru_cache(typed=True)` if it's going to exist.

nascheme replied:
You are right that it probably should live in `urllib.parse` somewhere.  I renamed it and improved the comment to make it a bit better.  Making it a function exported from `urllib` will take more thought yet.

In RFC 2616, it specifies that the request line has the syntax:
```
Request-URI    = "*" | absoluteURI | abs_path | authority
```
I think `http.server` is only handling the `abs_path` case.  Is it possible to pass absolute URIs to the server?

nascheme replied:
The redirection only happens if `http.server` is treating the request as a local file.  So I think my change is okay in only handling the `abs_path` case.

gpshead replied:
yeah i settled on only handling the abs path case in mine as well - effectively back to the PR the original reporter/author wrote as far as the code fix goes.  when it's got a scheme and netloc that passes through to Location just fine.

nascheme committed:
sanitize paths that can be confused as scheme

nascheme committed:
Add blurb.

nascheme committed:
Improve name and comment for _get_redirect_url().

nascheme committed:
Add urllib.parse.pathsplit() function.

nascheme committed:
Fix blurb formating, add note about pathsplit().

nascheme committed:
Improve comments and docs.

The pathsplit() function will work correctly on relative paths too so
don't say "absolute paths".  Improve comment for _get_redirect_url().

nascheme committed:
Add basic unit test for pathsplit().

nascheme committed:
Fix markup in blurb.

gpshead commented on Lib/http/server.py:
```diff
@@ -881,6 +878,23 @@ def guess_type(self, path):
+    # reported in gh-87389, a path starting with a double slash should not be
+    # treated as a relative URI.  Also, a path with a colon in the first
+    # component could also be parsed wrongly.
+    parts = urllib.parse.pathsplit(path)
```
Unfortunately BaseHTTPRequestHandler `self.path` isn't guaranteed to be only a pathname by anything.

`GET http://netloc/path/to/thing HTTP/1.0`

Is a valid HTTP query to many http servers.  Our existing http server code ignores the scheme and netloc on that and serves up /path/to/thing properly.  This code _might_ break things doing that?  ie: if http://netloc winds up in the returned path, the it seems like that'd move through pathsplit and urlunsplit properly only by chance rather than design just given the name and purpose of the pathsplit API vs what was just handed to it as input.

Check out my updated test in https://github.com/python/cpython/issues/87389.

I have no idea if anything depends on this as it should be unusual for any client not pretending to be talking to an HTTP proxy to make requests that way as that isn't normal per the http/1.0 and 1.1 RFCs but as it worked before and other servers like Apache continue to support it, I don't think we can break that - at least not in a bugfix.

nascheme replied:
Based on my reading of the RFC, `GET http://netloc/path/to/thing HTTP/1.0` should technically be allowed.  However, `http.server` doesn't handle it, before or after this PR.  Instead, it tries to use the second word as an absolute file path. I feel like we should not attempt to fix `http.server` to handle the above.  For HTTP 1.0, using only a file path was required.  For `curl` I notice that even with the `--http1.1` flag, it passes a path and not the URL with the scheme+netloc.

The bug causing #87389, IMHO, is that `http.server` was not consistent in how it was treating `self.path`.  The `translate_path()` method was treating it as a path and not a URL.  But, the redirect to add a trailing slash (done only if a folder with that name was found), parsed it as allowing a full URL.  It needs to do it one way or the other consistently.  The `_request_path_split()` method I added does this.  It doesn't make sense to add this to `urllib.parse` since what `http.server` is doing is not modern HTTP.

gpshead commented on Lib/urllib/parse.py:
```diff
@@ -480,6 +480,29 @@ def urlsplit(url, scheme='', allow_fragments=True):
+# typed=True avoids BytesWarnings being emitted during cache key
+# comparison since this API supports both bytes and str input.
+@functools.lru_cache(typed=True)
+def pathsplit(path):
```
while you've got a use for this, i'm wary about adding yet another public API to urllib.parse code.

nascheme replied:
Hmm, your previous comment suggested you wanted it in `urllib.parse`.  I'm wary of adding a new API as well. Now, I'm not even totally sure that having `http.server` use `pathsplit()` is correct at all.  RFC 2616 says:

> The most common form of Request-URI is that used to identify a
   resource on an origin server or gateway. In this case the absolute
   path of the URI MUST be transmitted (see [section 3.2.1](https://www.rfc-editor.org/rfc/rfc2616#section-3.2.1), abs_path) as
   the Request-URI, and the network location of the URI (authority) MUST
   be transmitted in a Host header field.

My current understanding is that `http.server` implements HTTP 1.0 and so it doesn't handle an absolute URI as Request-URI.  In that case, I think having `pathsplit()` is of pretty limited use since it probably is only useful if you are parsing an HTTP 1.0 request line.

nascheme replied:
I've removed `pathsplit()` as a public API.

gpshead replied:
The http1.1 rfc 2616 says it MUST be abs_path, not absoluteURI. Thus what we handle is correct (and we do have http/1.1 support IIRC).  If you follow the modern RFC 7231 i believe you end up in https://datatracker.ietf.org/doc/html/rfc3986#section-3.3

gpshead commented on Misc/NEWS.d/next/Security/2022-06-16-12-13-55.gh-issue-87389.MS9wAR.rst:
```diff
@@ -0,0 +1,8 @@
+the path as a net location or scheme.  Add :func:`urllib.parse.pathsplit`
+function.
+
+Co-authored-by: Gregory P. Smith <gps@google.com>
```
I try not to publish this email address lest people think they can contact me there. "Gregory P. Smith [Google]" is what I attempt to type when I'm conscious of things.

gpshead commented on Lib/test/test_urlparse.py:
```diff
@@ -1101,6 +1101,35 @@ def test_urlsplit_normalization(self):
+            (('', '', 'a:b', '', ''), './a:b'),
+        ]
+        for parts, result in cases:
+            self.assertEqual(urllib.parse.urlunsplit(parts), result)
```
add `, msg=f'{parts=}'` to these asserts so that a failing test makes it obvious what the inputs were.

gpshead commented on Lib/test/test_httpservers.py:
```diff
@@ -418,6 +418,26 @@ def test_undecodable_filename(self):
         self.check_status_and_reason(response, HTTPStatus.OK,
                                      data=os_helper.TESTFN_UNDECODABLE)
 
+    def test_get_dir_redirect_location_domain_injection_bug(self):
```
pull in my latest test.

nascheme committed:
Use pathsplit() in translate_path().

This avoids "manual parsing" of the Request-URI part of the request and
matches what _get_redirect_url() does.

nascheme committed:
Improved unit tests from gps.

nascheme committed:
Add test for Request-URI containing scheme.

nascheme committed:
Make _get_redirect_url() into a method.

Possible that someone could override this so a method is nicer.

nascheme committed:
Futher cleanups, remove urllib.parse.pathsplit().

Since pathsplit() doesn't seem like a generally useful public API,
remove it.  Instead, add a _request_path_split() method.  This ensures
that the redirect logic and the translate_path() method use the same
path parsing.

gpshead committed:
reword news

vadmium commented on Lib/urllib/parse.py:
```diff
@@ -491,14 +491,32 @@ def urlunparse(components):
 
+# Returns true if path can confused with a scheme.  I.e. a relative path
+# without leading dot that includes a colon in the first component.
+_is_scheme_like = re.compile(r'[^/.][^/]*:').match
```
Why the special allowance for a leading dot? Is there a test case for it? Yes, a scheme cannot start with a dot, but a _path-noscheme_ component of `.:` is no more legal than `https:` according to RFC 3986.

vadmium commented on Lib/test/test_urlparse.py:
```diff
@@ -1101,6 +1101,25 @@ def test_urlsplit_normalization(self):
+            # expected result is a relative URL without netloc and scheme
+            (('', 'a', '', '', ''), '//a'),
+            # extra leading slashes need to be stripped to avoid confusion
+            # with a relative URL
```
confusion with a _protocol-relative_ URL? [as opposed to a host-relative URL]

vadmium reviewed (commented):
There seem to be three behaviour changes proposed:

1. Changing urlunsplit(('', '', '//path', '', '')) to return '/path' instead of '//path'

Another option is to prefix with double slash, representing an empty host e.g. '////path'. This is proposed in issue #78457 and pull request #113563. I think I prefer that four-slash option, because it also fixes some _urlsplit_ → _urlunsplit_ round-trip cases.

2. Changing urlunsplit(('', '', 'colon:path', '', '')) → './colon:path'

This seems a reasonable change, and it is kind of suggested in RFC 3986. (Another option might be to encode the first colon, and return 'colon%3Apath'.)

3. SimpleHTTPRequestHandler’s handling of `GET https://example.net/dir`

This is a legal HTTP 1.0 and 1.1 request, but is mainly for proxy servers, which is not what SimpleHTTPRequestHandler does. In this case the server looks up https:/example.net/dir as a path in its filesystem (which is not in spirit of HTTP), and decides to redirect with a trailing slash.

Currently it looks like the code sends `Location: https://example.net/dir/`. I don’t think there is anything really wrong with that.

The proposed changes would send `Location: ./https://example.net/dir/`. This new redirect is a path-relative URL. The base URL is supposed to be the original target `https://example.net/dir`, so the redirect would resolve to `https://example.net/https://example.net/dir/`, which is not intended.

If you want to fix anything in the HTTP server, I would make the server ignore the scheme and authority components, and just look up the path component. But I don’t think anyone is complaining about that, so it may not be worth fixing.

If the _urllib.parse_ changes are too disruptive, perhaps a deprecation warning is the best way forward, and either add an opt-in way to get the new behaviour, or change the warning to an exception in the future?

serhiy-storchaka committed:
Merge branch 'main' into urlunsplit_relative

serhiy-storchaka committed:
Merge branch 'main' into urlunsplit_relative
